File: crm/src/main/java/serp/project/crm/ui/controller/ContactController.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.crm.ui.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import serp.project.crm.core.domain.dto.PageRequest;
import serp.project.crm.core.domain.dto.request.CreateContactRequest;
import serp.project.crm.core.domain.dto.request.UpdateContactRequest;
import serp.project.crm.core.usecase.ContactUseCase;
import serp.project.crm.kernel.utils.AuthUtils;

@RestController
@RequestMapping("/api/v1")
@RequiredArgsConstructor
@Slf4j
public class ContactController {

    private final ContactUseCase contactUseCase;
    private final AuthUtils authUtils;

    @PostMapping("/customers/{customerId}/contacts")
    public ResponseEntity<?> createContact(
            @PathVariable Long customerId,
            @Valid @RequestBody CreateContactRequest request) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        Long userId = authUtils.getCurrentUserId().orElse(null);
        if (tenantId == null || userId == null) {
            return null;
        }

        log.info("POST /api/v1/customers/{}/contacts - Creating contact for tenant: {}", customerId, tenantId);
        request.setCustomerId(customerId);
        var response = contactUseCase.createContact(request, userId, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PatchMapping("/customers/{customerId}/contacts/{id}")
    public ResponseEntity<?> updateContact(
            @PathVariable Long customerId,
            @PathVariable Long id,
            @Valid @RequestBody UpdateContactRequest request) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        Long userId = authUtils.getCurrentUserId().orElse(null);
        if (tenantId == null || userId == null) {
            return null;
        }
        log.info("PATCH /api/v1/customers/{}/contacts/{} - Updating contact for tenant: {}", customerId, id, tenantId);
        var response = contactUseCase.updateContact(id, request, userId, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/customers/{customerId}/contacts/{id}")
    public ResponseEntity<?> getContactById(
            @PathVariable Long customerId,
            @PathVariable Long id) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        if (tenantId == null) {
            return null;
        }

        log.info("GET /api/v1/customers/{}/contacts/{} - Fetching contact for tenant: {}", customerId, id, tenantId);
        var response = contactUseCase.getContactById(id, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/customers/{customerId}/contacts")
    public ResponseEntity<?> getAllContactsByCustomer(
            @PathVariable Long customerId) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        if (tenantId == null) {
            return null;
        }

        log.info("GET /api/v1/customers/{}/contacts - Fetching all contacts for tenant: {}", customerId, tenantId);

        var response = contactUseCase.getContactsByCustomerId(customerId, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @DeleteMapping("/customers/{customerId}/contacts/{id}")
    public ResponseEntity<?> deleteContact(
            @PathVariable Long customerId,
            @PathVariable Long id) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);
        if (tenantId == null) {
            return null;
        }

        log.info("DELETE /api/v1/customers/{}/contacts/{} - Deleting contact for tenant: {}", customerId, id, tenantId);
        var response = contactUseCase.deleteContact(id, tenantId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/contacts")
    public ResponseEntity<?> getAllContacts(
            @RequestParam(defaultValue = "1") Integer page,
            @RequestParam(defaultValue = "20") Integer size) {
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);

        log.info("GET /api/v1/contacts - Fetching all contacts for tenant: {}, page: {}, size: {}", tenantId,
                page, size);

        PageRequest pageRequest = PageRequest.builder()
                .page(page)
                .size(size)
                .build();

        var response = contactUseCase.getAllContacts(tenantId, pageRequest);
        return ResponseEntity.status(response.getCode()).body(response);
    }
}

File: api_gateway/src/ui/controller/crm/contact_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	"github.com/serp/api-gateway/src/core/service/crm"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type ContactController struct {
	contactService crm.IContactService
}

func (c *ContactController) CreateContact(ctx *gin.Context) {
	customerId, ok := utils.ValidateAndParseID(ctx, "customerId")
	if !ok {
		return
	}

	var payload map[string]any
	if err := ctx.ShouldBindJSON(&payload); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	payload["customerId"] = customerId

	res, err := c.contactService.CreateContact(ctx.Request.Context(), customerId, payload)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (c *ContactController) UpdateContact(ctx *gin.Context) {
	customerId, ok := utils.ValidateAndParseID(ctx, "customerId")
	if !ok {
		return
	}

	contactId, ok := utils.ValidateAndParseID(ctx, "contactId")
	if !ok {
		return
	}

	var payload map[string]any
	if err := ctx.ShouldBindJSON(&payload); err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralBadRequest)
		return
	}

	payload["customerId"] = customerId
	payload["id"] = contactId

	res, err := c.contactService.UpdateContact(ctx.Request.Context(), customerId, contactId, payload)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (c *ContactController) GetContactByID(ctx *gin.Context) {
	customerId, ok := utils.ValidateAndParseID(ctx, "customerId")
	if !ok {
		return
	}

	contactId, ok := utils.ValidateAndParseID(ctx, "contactId")
	if !ok {
		return
	}

	res, err := c.contactService.GetContactByID(ctx.Request.Context(), customerId, contactId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (c *ContactController) GetContactsByCustomer(ctx *gin.Context) {
	customerId, ok := utils.ValidateAndParseID(ctx, "customerId")
	if !ok {
		return
	}

	res, err := c.contactService.GetContactsByCustomer(ctx.Request.Context(), customerId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (c *ContactController) GetContacts(ctx *gin.Context) {
	page, size, valid := utils.ValidatePaginationParams(ctx)
	if !valid {
		return
	}

	res, err := c.contactService.GetContacts(ctx.Request.Context(), page, size)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func (c *ContactController) DeleteContact(ctx *gin.Context) {
	customerId, ok := utils.ValidateAndParseID(ctx, "customerId")
	if !ok {
		return
	}

	contactId, ok := utils.ValidateAndParseID(ctx, "contactId")
	if !ok {
		return
	}

	res, err := c.contactService.DeleteContact(ctx.Request.Context(), customerId, contactId)
	if err != nil {
		utils.AbortErrorHandle(ctx, constant.GeneralInternalServerError)
		return
	}
	ctx.JSON(res.Code, res)
}

func NewContactController(contactService crm.IContactService) *ContactController {
	return &ContactController{contactService: contactService}
}
